cogs/heartbeat/testCore.py

```diff
@@ -0,0 +1,73 @@
+import asyncio
+from contextlib import asynccontextmanager
+import logging
+from typing import Union
+from unittest import mock
+
+import aiohttp
+import pytest
+from pytest import LogCaptureFixture, MonkeyPatch
+
+from redbot.core.bot import Red
+
+from . import constants, core
+from .heartbeat import Heartbeat
+
+
+def createMockClientResponseGet(response: aiohttp.ClientResponse):
+    @asynccontextmanager
+    async def mockGet(*_args, **_kwargs):
+        yield response
+
+    return mockGet
+
+
+@pytest.mark.asyncio
+async def testLoopBad(caplog: LogCaptureFixture, monkeypatch: MonkeyPatch, red: Red):
+    """Test to ensure `_loop` works as expected with bad case: the response from the push URL is always non-OK."""
+
+    # prep
+    red.loop = mock.create_autospec(spec=asyncio.AbstractEventLoop)
+    minInterval = 1
+    maxFailedPings = 3
+    core.MAX_FAILED_PINGS = maxFailedPings
+
+    # mock
+    mockResponse: Union[mock.Mock, aiohttp.ClientResponse] = mock.create_autospec(
+        spec=aiohttp.ClientResponse,
+        status=500,  # i.e., HTTP 500
+    )
+
+    # patch
+    monkeypatch.setattr(
+        target=aiohttp.ClientSession,
+        name="get",
+        value=createMockClientResponseGet(response=mockResponse),
+    )
+
+    # log
+    caplog.set_level(level=logging.ERROR)
+
+    # cog
+    await red.add_cog(Heartbeat(bot=red))
+    cog: Heartbeat = red.get_cog("Heartbeat")
+    assert cog and isinstance(cog, Heartbeat)
+
+    # prep
+    testInterval = minInterval
+    await cog.config.get_attr(constants.KEY_INSTANCE_NAME).set("test instance")
+    await cog.config.get_attr(constants.KEY_PUSH_URL).set("test URL")
+    await cog.config.get_attr(constants.KEY_INTERVAL).set(testInterval)
+
+    # test
+    try:
+        await asyncio.wait_for(cog._loop(), testInterval * core.MAX_FAILED_PINGS * 2)
```

Comment:
Is there anyway to simulate this time going forward rather than actually waiting for this time to pass in real time?

Reply:
We might be able to get around by doing something hacky with the `asyncio.sleep` calls in `Heartbeat._loop` (e.g., increment a variable counting the amount of time slept, assuming the time spent by other operations is negligible), but other than that I don't think we can _simulate_ the duration, because the `_loop` itself is not duration-based. It sleeps as it executes.

Alternatively, because this test case tests a bad case where the loop *must* fail, we can totally intentionally manipulate the effect of `asyncio.sleep` calls (e.g., by mock-patching) to avoid waiting in real time, and test for similar criteria (i.e., the loop should have been done, and the error log line should be as expected).

Reply:
I'm fine with resolving as is, but just something to keep in mind for any sleep-based loops we have to test in the future.

Reply:
Updated diff makes sense to me, resolving.